pulpcore/app/replica.py

```diff
@@ -0,0 +1,206 @@
+from django.db.models import Model
+
+from pulp_glue.common.context import PulpContext
+from pulpcore.tasking.tasks import dispatch
+from pulpcore.app.tasks.base import general_update, general_create, general_delete
+from pulpcore.plugin.util import get_url
+
+
+class ReplicaContext(PulpContext):
+    def prompt(self, *args, **kwargs):
+        pass
+
+    def echo(self, *args, **kwargs):
+        pass
+
+
+class Replicator:
+    distribution_ctx_cls = None
+    repository_ctx_cls = None
+    publication_ctx_cls = None
+    remote_model_cls = None
+    repository_model_cls = None
+    distribution_model_cls = None
+    distribution_serializer_name = None
+    repository_serializer_name = None
+    remote_serializer_name = None
+    app_label = None
+    sync_task = None
+
+    def __init__(self, pulp_ctx, task_group):
+        """
+        :param pulp_ctx: PulpReplicaContext
+        :param task_group: TaskGroup
+        """
+        self.pulp_ctx = pulp_ctx
+        self.task_group = task_group
+
+    @staticmethod
+    def needs_update(fields_dict, model_instance):
+        """
+        Compares a Model instance's attributes against a dictionary where keys are attribute
+        names and values are expected values.
+        """
+        needs_update = False
+        for field_name, value in fields_dict.items():
+            if isinstance(getattr(model_instance, field_name), Model):
+                if get_url(getattr(model_instance, field_name)) != value:
+                    needs_update = True
+            elif getattr(model_instance, field_name) != value:
+                needs_update = True
+        return needs_update
+
+    def get_upstream_distributions(self, labels=None):
```

Comment:
```suggestion
    def upstream_distributions(self, labels=None):
```
As being an iterator this can be considered to _be_ the upstream distributions.